core/node/base_token_adjuster/src/base_token_l1_behaviour.rs

```diff
@@ -0,0 +1,331 @@
+use std::{
+    cmp::max,
+    ops::{Div, Mul},
+    sync::Arc,
+    time::Instant,
+};
+
+use anyhow::Context;
+use bigdecimal::{num_bigint::ToBigInt, BigDecimal, Zero};
+use zksync_config::BaseTokenAdjusterConfig;
+use zksync_eth_client::{BoundEthInterface, CallFunctionArgs, Options};
+use zksync_node_fee_model::l1_gas_price::TxParamsProvider;
+use zksync_types::{
+    base_token_ratio::BaseTokenAPIRatio,
+    ethabi::{Contract, Token},
+    web3::{contract::Tokenize, BlockNumber},
+    Address, U256,
+};
+
+use crate::metrics::{OperationResult, OperationResultLabels, METRICS};
+
+#[derive(Debug, Clone)]
+pub struct UpdateOnL1Params {
+    pub eth_client: Box<dyn BoundEthInterface>,
+    pub gas_adjuster: Arc<dyn TxParamsProvider>,
+    pub token_multiplier_setter_account_address: Address,
+    pub chain_admin_contract: Contract,
+    pub getters_facet_contract: Contract,
+    pub diamond_proxy_contract_address: Address,
+    pub chain_admin_contract_address: Option<Address>,
+    pub config: BaseTokenAdjusterConfig,
+}
+
+#[derive(Debug, Clone)]
+pub enum BaseTokenL1Behaviour {
+    UpdateOnL1 {
+        params: UpdateOnL1Params,
+        last_persisted_l1_ratio: Option<BigDecimal>,
+    },
+    NoOp,
+}
+
+impl BaseTokenL1Behaviour {
+    pub async fn update_l1(&mut self, new_ratio: BaseTokenAPIRatio) -> anyhow::Result<()> {
+        let (l1_params, last_persisted_l1_ratio) = match self {
+            BaseTokenL1Behaviour::UpdateOnL1 {
+                ref params,
+                ref last_persisted_l1_ratio,
+            } => (&params.clone(), last_persisted_l1_ratio),
+            BaseTokenL1Behaviour::NoOp => return Ok(()),
+        };
+
+        let prev_ratio = if let Some(prev_ratio) = last_persisted_l1_ratio {
+            prev_ratio.clone()
+        } else {
+            let prev_ratio = self.get_current_ratio_from_l1(l1_params).await?.clone();
+            self.update_last_persisted_l1_ratio(prev_ratio.clone());
+            tracing::info!(
+                "Fetched current base token ratio from the L1: {}",
+                prev_ratio.to_bigint().unwrap()
+            );
+            prev_ratio.clone()
```

Comment:
nit: looks like first and third clones and needless, not sure why lint check doesn't catch it

Reply:
Good catch - fixed